### PR TITLE
Pass full input_shape to cell.build()

### PR DESCRIPTION
**Pull request by OverLordGoldDragon**
`step_input_shape` unnecessarily deprives `Cell` instances of potentially necessary information - namely, custom architectures may require the layer's `timesteps` at build time (before `call()`). There are no performance or memory gains from trimming a list by one, nor really in clarity as this isn't being passed to `call` where the input shape does omit the time dimension (and in that case `input_shape` as opposed to `step_input_shape` is misleading anyway).

This commit does not require changing any other code in `recurrent` or `recurrent_v2`, as `input_shape` is indexed via `[-1]` - though unsure for the rest of the repository; either way, nothing is lost with this commit.

### CODE REVIEW

**Comment by qlzh727** [comment]
This breaks the API contract that layers.build() method will get the shape of input tensor, which should be same as the one passed in when call() is invoked. So far the cell only rely on the last dim of the inputs to build the weights, but this might not be true for other customized cells, eg if the cell somehow rely on the batch size to do certain behavior, in the time_major input case, the first dim is no longer batch, but timestep if we pass the full input shape.

**Comment by OverLordGoldDragon** [comment]
@qlzh727 Is [recurrent_v2](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/layers/recurrent_v2.py#L1118) exempt from this contract, then? `timesteps` is extracted from `input_shape`, so its `call()` has a different `input_shape` than its `build()` (via `super()`).

**Comment by qlzh727** [comment]
I am not sure what you mean. In LSTM v2, the call() and build() will get same input_shape. Could u be more specific about the different shape here?

**Comment by OverLordGoldDragon** [comment]
@qlzh727 `input_shape` isn't actually passed in, but extracted as [`K.int_shape(inputs)`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/layers/recurrent_v2.py#L1118), and the layer doesn't define its own `build()`,  but inherits it from `recurrent.LSTM` - so basically a "trick" to circumvent the contract. Well, such tricks are quite inconvenient to implement for custom RNN layers - one's via re-implementing the base class (`RNN`), another is to pass `timesteps` to each layer in the model by fetching it manually from preceding layers.

I'm guessing regardless of who oversees this 'contract', the process for possibly changing it will take months - so instead, how about this: pass full `input_shape` to _both_, `build()` and `call()` of all recurrent layers. Would that work?

**Comment by qlzh727** [comment]
> I'm guessing regardless of who oversees this 'contract', the process for possibly changing it will take months - so instead, how about this: pass full input_shape to both, build() and call() of all recurrent layers. Would that work?

All the recurrent layer IS currently getting full input_shape for both build() and call(). All the recurrent cells is getting input shapes without the timestep dimension. I don't think this contract need to be changed for RNN.

My question, however, is why the cell need timestep in the build() method. A cell should handle the input one timestep by one timestep, and it shouldn't care how many timesteps there will be for the whole sequence. Can u be more specific about why the timestep is needed? (Is it related to https://github.com/tensorflow/tensorflow/issues/36559?)

**Comment by OverLordGoldDragon** [comment]
@qlzh727 I include the time dimension in the "full" input shape, as that's what the layer is fed in the model network. And you are correct, this is about recurrent batch normalization - but also [IndRNNs](https://arxiv.org/abs/1803.04831), which require `timesteps` at build time for `recurrent_kernel` initialization. I've implemented the layer in `tf.keras` in TF 1 & 2 without publishing - and for the record, it easily dominates LSTMs and GRUs in my application, and I may PR it in the future.

This said, while TF currently doesn't have any layers that need `timesteps` at build time, some users (like myself) may need it for custom functionality. Nothing is lost from passing it instead of "step input shape" - so omitting the time dimension is unnecessarily prohibitive.

**Comment by qlzh727** [comment]
Sorry for the late reply. If your cell need to be timestep aware, then it is no longer a cell, but a RNN layer, since the cell only exposed to one timestep at a time. Maybe you should create subclass RNN layer instead of a cell?

**Comment by OverLordGoldDragon** [comment]
@qlzh727 `build()` belongs to cells, and `RNN` base class never passes the full input shape to either the layer or the cell. Hence, since all RNNs inherit `RNN` - either (1) `RNN` must be re-implemented, or (2) `timesteps` must be passed to the layer as an argument. Both add much redundancy.

`call()` could also be `timesteps`-dependent, and _is_ for recurrent batchnorm; at inference time, generalizing to `timesteps` longer than trained uses the last train step's moving average.

Can you suggest an 'API-friendly' means of injecting `timesteps` information into a cell at build time? If so, I'll consider this resolved - but whatever I thought of so far is overly hackish.

**Comment by qlzh727** [comment]
I would suggest you to sub class RNN layer for your normalization case.

1. Even we pass the full input_shape from RNN layer to cell.build(), it is also very likely that the timestep dimention is None. It is very common for user to build a model with dynamic timestep size. I don't think you can build weights with None as any part of the shape.

2. In the case of normalization, eg, you will have mean and variance for the batch of the data in your RNN layer, why those value has to aware of the length of timestep? especially at each individual step? I would expect the mean and variance are calculate against the feature dim (which is the last axis, rather than the timestep axis).

3. For IndRNNs, I check the existing implementation in TF contrib. The IndRNNCell also build with (batch, feature) as input shape,  and doesn't need timestep. The code was in contrib, which is not visible on HEAD, but you can get it from TF 1.15 release. Here is some snippet from the it:

```
class IndRNNCell(rnn_cell_impl.LayerRNNCell):
  """Independently Recurrent Neural Network (IndRNN) cell
    (cf. https://arxiv.org/abs/1803.04831).

  Args:
    num_units: int, The number of units in the RNN cell.
    activation: Nonlinearity to use.  Default: `tanh`.
    reuse: (optional) Python boolean describing whether to reuse variables
     in an existing scope.  If not `True`, and the existing scope already has
     the given variables, an error is raised.
    name: String, the name of the layer. Layers with the same name will
      share weights, but to avoid mistakes we require reuse=True in such
      cases.
    dtype: Default dtype of the layer (default of `None` means use the type
      of the first input). Required when `build` is called before `call`.
  """

  def __init__(self,
               num_units,
               activation=None,
               reuse=None,
               name=None,
               dtype=None):
    super(IndRNNCell, self).__init__(_reuse=reuse, name=name, dtype=dtype)

    # Inputs must be 2-dimensional.
    self.input_spec = input_spec.InputSpec(ndim=2)

    self._num_units = num_units
    self._activation = activation or math_ops.tanh

  @property
  def state_size(self):
    return self._num_units

  @property
  def output_size(self):
    return self._num_units

  def build(self, inputs_shape):
    if tensor_shape.dimension_value(inputs_shape[1]) is None:
      raise ValueError(
          "Expected inputs.shape[-1] to be known, saw shape: %s" % inputs_shape)

    input_depth = tensor_shape.dimension_value(inputs_shape[1])
    # pylint: disable=protected-access
    self._kernel_w = self.add_variable(
        "%s_w" % rnn_cell_impl._WEIGHTS_VARIABLE_NAME,
        shape=[input_depth, self._num_units])
    self._kernel_u = self.add_variable(
        "%s_u" % rnn_cell_impl._WEIGHTS_VARIABLE_NAME,
        shape=[1, self._num_units],
        initializer=init_ops.random_uniform_initializer(
            minval=-1, maxval=1, dtype=self.dtype))
    self._bias = self.add_variable(
        rnn_cell_impl._BIAS_VARIABLE_NAME,
        shape=[self._num_units],
        initializer=init_ops.zeros_initializer(dtype=self.dtype))
    # pylint: enable=protected-access

    self.built = True

  def call(self, inputs, state):
    """IndRNN: output = new_state = act(W * input + u * state + B)."""

    gate_inputs = math_ops.matmul(inputs, self._kernel_w) + (
        state * self._kernel_u)
    gate_inputs = nn_ops.bias_add(gate_inputs, self._bias)
    output = self._activation(gate_inputs)
    return output, output
```

**Comment by OverLordGoldDragon** [comment]
@qlzh727 Thanks for the suggestions.

 1. It is even more likely for `batch_size` to be `None`, yet we require it for `stateful=True`; personally I've yet to see a `None` for `timesteps`, but just as for `stateful`, we can require `timesteps` to be defined for recurrent batchnorm.
 2. I'll clarify; the bottom of page 3 of the [paper](https://arxiv.org/abs/1603.09025) reads: 

> ... we estimate the population statistics separately for each timestep `1, ..., Tmax`, where `Tmax` is the length of the longest training sequence. When at test time we need to generalize beyond `Tmax`, we use the population statistics of time `Tmax` for all time steps beyond it.

Thus, the population statistics variables will be shaped `(Tmax, units)` - requiring the knowledge of `Tmax` _at build time_.

 3. The implementation is flawed; authors note the need for careful weight clipping for a certain range of timesteps to avoid exploding gradients - and is shown in the [official implementation](https://github.com/Sunnydreamrain/IndRNN_Theano_Lasagne/blob/master/adding/adding.py#L51) (alias `seq_len`). That implementation uses a safe default for several cases, but not all (and `_kernel_w` also lacks an initializer).

**Comment by qlzh727** [comment]
If the intention for this PR is trying to address your BN RNN issue, then I would suggest take the subclass RNN layer approach, but this will require some work.

1. In your BN RNN layer, each step has to aware that what the current step number is, and find the corresponding BN variable to use. The current RNN API doesn't pass any step number information to the cell, so finding the corresponding BN variable will be tricky. The implementation in your https://github.com/tensorflow/tensorflow/issues/36559 might not be correct since the step function is called once first to calculate the output shape when creating the output TensorArray. The result of that is ignored. 

2. The step information is available in the loop function in the keras backend.rnn() while iterating through the time sequence, we don't pass that info to the cell since the cell doesn't need to aware of it. In you case, you probably want to create a BN variable in your RNN layer. At the time when calling the cell, you need to reconstruct you input into a nested structure, with the BN variable. The backend.rnn should handle the nested input correctly by sliding on the BN with timestep dim, same as the input. You cell should get real input and the slide of BN variable in the call() as the first parameter, and you might want to return the value for the BN variable in the result as well.

3. At the end of BN RNN layer call() body, you probably want to gather the result of BN and add update to the original variable. Your cell probably shouldn't worry about updating the BN since they are just using the value. 

With that I think your cell don't need to worry about the timestep information since it really shouldn't, and the layer will handle it.

**Comment by OverLordGoldDragon** [comment]
@qlzh727 I suppose I haven't sufficiently considered a `_v2`-like approach, utilizing `K.rnn` directly via the layer's `call()`; since it recieves the complete input, I should be able to build variables at call time, with `if not self.bn_built`-like logic. I've just been trying to stick to `tf.keras` layer structuring, as I presume it's more stable and optimized - but this does seem fitting. You are also correct regarding `step` - I've re-implemented it since and confirmed it works via `tf.print`.

The intent wasn't just to get RBN working, but that the current policy adds redundancy that's easy to change. Regardless, I've suspended my works on RBN per performance and correctness [concerns](https://github.com/tensorflow/tensorflow/issues/36797); perhaps my mistake is trying to use the exact loop function that `tf.keras` uses, where something else might be more appropriate. I may look into it again in some future, only if TF Profiler's fixed in 2.1+.

In any case, though I've yet to try it, your approach seems promising, and I don't have much else to add on the issue for now - so I'll close the PR. Also, thanks for following through with this, here and on other threads - you are the most responsive TF developer I've seen.
